Add network docs (#35)
NeurodataWithoutBorders/nwb_benchmarks: docs/writing_benchmarks.rst

        nwbfile = io.read()
        return (nwbfile, io, file, byte_stream)

and so we managed to save ~5 lines of code for every occurrence of this logic in the benchmarks. Good choices of function names are critical to effectively communicating the actions being undertaken. Thorough annotation of signatures is likewise critical to understanding input/output relationships for these functions.

.. _network-tracking-benchmarks:

Writing a network tracking benchmark
------------------------------------

> Functions that require network access ---such as reading a file from S3--- are often a black box, with functions in other libraries (e.g., `h5py`, `fsspec` etc.) managing the access to the remote resources. The runtime performance of such functions is often inherently driven by how these functions utilize the network to access the resources. It is, hence, important that we can profile the network traffic that is being generated to better understand, e.g., the amount of data that is being downloaded and uploaded, the number of requests that are being sent/received, and others.

Reviewer: ```suggestion
Functions that require network access ---such as reading a file from S3--- are often a black box, with functions in other libraries (e.g., ``h5py``, ``fsspec`` etc.) managing the access to the remote resources. The runtime performance of such functions is often inherently driven by how these functions utilize the network to access the resources. It is, hence, important that we can profile the network traffic that is being generated to better understand, e.g., the amount of data that is being downloaded and uploaded, the number of requests that are being sent/received, and others.
```

> Functions that require network access ---such as reading a file from S3--- are often a black box, with functions in other libraries (e.g., `h5py`, `fsspec` etc.) managing the access to the remote resources. The runtime performance of such functions is often inherently driven by how these functions utilize the network to access the resources. It is, hence, important that we can profile the network traffic that is being generated to better understand, e.g., the amount of data that is being downloaded and uploaded, the number of requests that are being sent/received, and others.

Reviewer: ```suggestion
Functions that require network access ---such as reading a file from S3--- are often a black box, with functions in other libraries (e.g., `h5py`, `fsspec` etc.) managing the access to the remote resources. The runtime performance of such functions is often inherently driven by how these functions utilize the network to access the resources. It is, hence, important that we can profile the network traffic that is being generated to better understand the amount of data that is being downloaded and uploaded, the number of requests that are being sent/received, and so on.
```

> To simplify the implementation of benchmarks for tracking network statistics, we implemented in the `nwb_benchmarks.core` module various helper classes and functions. The network tracking functionality is designed to track the network traffic generated by the main Python process that our tests are running during a user-defined period of time. The `network_activity_tracker` context manager can be used to track the network traffic generated by the code within the context. A basic network benchmark, then looks as follows:

Reviewer: ```suggestion
To simplify the implementation of benchmarks for tracking network statistics, we implemented in the ``nwb_benchmarks.core`` module various helper classes and functions. The network tracking functionality is designed to track the network traffic generated by the main Python process that our tests are running during a user-defined period of time. The ``network_activity_tracker`` context manager can be used to track the network traffic generated by the code within the context. A basic network benchmark, then looks as follows:
```

.. code-block:: python

    from nwb_benchmarks import TSHARK_PATH
    from nwb_benchmarks.core import network_activity_tracker
    import requests   # Only used here for illustration purposes

    class SimpleNetworkBenchmark:

        def track_network_activity_uri_request():
            with network_activity_tracker(tshark_path=TSHARK_PATH) as network_tracker:
                x = requests.get('https://nwb-benchmarks.readthedocs.io/en/latest/setup.html')
            return network_tracker.asv_network_statistics

> In cases where a context manager may not be sufficient, we can alternatively use the `NetworkTracker` class directly to explicitly control when to start and stop the tracking.

Reviewer: ```suggestion
In cases where a context manager may not be sufficient, we can alternatively use the ``NetworkTracker`` class directly to explicitly control when to start and stop the tracking.
```

.. code-block:: python

    from nwb_benchmarks import TSHARK_PATH
    from nwb_benchmarks.core import NetworkTracker
    import requests   # Only used here for illustration purposes

    class SimpleNetworkBenchmark:

        def track_network_activity_uri_request():
            tracker = NetworkTracker()
            tracker.start_network_capture(tshark_path=TSHARK_PATH)
            x = requests.get('https://nwb-benchmarks.readthedocs.io/en/latest/setup.html')
            tracker.stop_network_capture()
            return tracker.asv_network_statistics

> By default, the `NetworkTracker` and `network_activity_tracker` track the network activity of the current process ID (i.e., ``os.getpid()``), but the PID to track can also be set explicitly if a different process needs to be monitored.

Reviewer: ```suggestion
By default, the ``NetworkTracker`` and ``network_activity_tracker`` track the network activity of the current process ID (i.e., ``os.getpid()``), but the PID to track can also be set explicitly if a different process needs to be monitored.
```
